icu4c/source/i18n/timezone.cpp

```diff
@@ -1369,123 +1370,102 @@ TimeZone::getCustomID(const UnicodeString& id, UnicodeString& normalized, UError
 UBool
 TimeZone::parseCustomID(const UnicodeString& id, int32_t& sign,
                         int32_t& hour, int32_t& min, int32_t& sec) {
-    static const int32_t         kParseFailed = -99999;
-
-    NumberFormat* numberFormat = 0;
-    UnicodeString idUppercase = id;
-    idUppercase.toUpper("");
-
-    if (id.length() > GMT_ID_LENGTH &&
-        idUppercase.startsWith(GMT_ID, GMT_ID_LENGTH))
-    {
-        ParsePosition pos(GMT_ID_LENGTH);
-        sign = 1;
-        hour = 0;
-        min = 0;
-        sec = 0;
-
-        if (id[pos.getIndex()] == MINUS /*'-'*/) {
-            sign = -1;
-        } else if (id[pos.getIndex()] != PLUS /*'+'*/) {
+    if (id.length() < GMT_ID_LENGTH) {
+      return false;
+    }
+    if (0 != u_strncasecmp(id.getBuffer(), GMT_ID, GMT_ID_LENGTH, 0)) {
+        return false;
+    }
+    // ICU_Utility::parseNumber also accept non ASCII digits so we need to first
+    // check we only have ASCII chars.
+    if (!uprv_isInvariantUString(id.getBuffer(), id.length())) {
+        return false;
+    }
+    sign = 1;
+    hour = 0;
+    min = 0;
+    sec = 0;
+
+    if (id[GMT_ID_LENGTH] == MINUS /*'-'*/) {
+        sign = -1;
+    } else if (id[GMT_ID_LENGTH] != PLUS /*'+'*/) {
+        return false;
+    }
+
+    // Support the following cases
```

Comment:
You should move the comment for patterns without `:` into if block.

Reply:
done